zerver/lib/actions.py

```diff
@@ -4153,39 +4153,56 @@ def do_mark_stream_messages_as_read(user_profile: UserProfile,
     do_increment_logging_stat(user_profile, COUNT_STATS['messages_read_interactions::hour'],
                               None, event_time, increment=min(1, count))
     return count
 
-def do_clear_mobile_push_notifications_for_ids(user_profile: UserProfile,
+def do_update_mobile_push_notification(message: Message, prior_mention_user_ids: Set[int]) -> None:
+    # This is an initial implementation of this method.
+    # It was written primarily to fix https://github.com/zulip/zulip/pull/15428.
+    # Future implementations should include updating the messsage in the
+    # notification of the user after the message was edited.
+    # We don't need to care about wildcard mentions here since,
+    # we don't send notifications for users who muted the stream and
+    # were wildcard mentioned.
+    if not message.is_stream_message():
+        return
+
+    # remove notification for users who were no longer mentioned.
+    remove_notify_users = prior_mention_user_ids - message.mentions_user_ids
+    do_clear_mobile_push_notifications_for_ids(list(remove_notify_users), [message.id])
+
+def do_clear_mobile_push_notifications_for_ids(user_profile_ids: List[int],
                                                message_ids: List[int]) -> None:
     filtered_message_ids = list(UserMessage.objects.filter(
         message_id__in=message_ids,
-        user_profile=user_profile,
+        user_profile_id__in=user_profile_ids,
     ).extra(
         where=[UserMessage.where_active_push_notification()],
     ).values_list('message_id', flat=True))
 
     num_detached = settings.MAX_UNBATCHED_REMOVE_NOTIFICATIONS - 1
-    for message_id in filtered_message_ids[:num_detached]:
-        # Older clients (all clients older than 2019-02-13) will only
-        # see the first message ID in a given notification-message.
-        # To help them out, send a few of these separately.
-        queue_json_publish("missedmessage_mobile_notifications", {
-            "type": "remove",
-            "user_profile_id": user_profile.id,
-            "message_ids": [message_id],
-        })
+    for user_profile_id in user_profile_ids:
+        for message_id in filtered_message_ids[:num_detached]:
+            # Older clients (all clients older than 2019-02-13) will only
+            # see the first message ID in a given notification-message.
+            # To help them out, send a few of these separately.
+            queue_json_publish("missedmessage_mobile_notifications", {
+                "type": "remove",
+                "user_profile_id": user_profile_id,
+                "message_ids": [message_id],
+            })
     if filtered_message_ids[num_detached:]:
-        queue_json_publish("missedmessage_mobile_notifications", {
-            "type": "remove",
-            "user_profile_id": user_profile.id,
-            "message_ids": filtered_message_ids[num_detached:],
-        })
+        for user_profile_id in user_profile_ids:
```

Comment:
I don't think there is a way to send this as a bulk process since we need to push notify each user separately.

Reply:
Yeah, I think we can only do the database query in bulk (and we'll need to return the `(user_id, message_id)` pairs and do some dictionary work to organize things).  Which is still meaningful, I think, since then the loops we have will just be looping with `send_event` over users who _actually_ got a push notification, rather than doing 1 query per subscriber.

Reply:
I guess the difference isn't that meaningful, since the caller knows all the users it passes were mentioned; but it might be relevant in future worlds where we call this in more settings.

Reply:
I think we should just make to improve upon this next time someone (possibly me) works on this system.

Reply:
I don't think it's correct as written; right now, in addition to this function being a land mine that would need a `assert len(user_profile_ids) == 1 or len(message_ids) == 1` to avoid being called improperly, I noticed we'll send a "remove" notification for every passed user for the removed message regardless of whether that specific user should.

We need to return `(user_id, message_id)` pairs and process them properly.  Something like this:
```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 2044e3ca7a..726b47d932 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -4183,15 +4183,23 @@ def do_update_mobile_push_notification(message: Message, prior_mention_user_ids:
 
 def do_clear_mobile_push_notifications_for_ids(user_profile_ids: List[int],
                                                message_ids: List[int]) -> None:
-    filtered_message_ids = list(UserMessage.objects.filter(
+    # This functions supports clearing notifications for several users
+    # only for the message-edit use case where we'll have a single user_id.
+    assert len(user_profile_ids) == 1 or len(message_ids) == 1
+
+    messages_by_user = defaultdict(list)
+    notifications_to_update = list(UserMessage.objects.filter(
         message_id__in=message_ids,
         user_profile_id__in=user_profile_ids,
     ).extra(
         where=[UserMessage.where_active_push_notification()],
-    ).values_list('message_id', flat=True))
+    ).values_list('user_profile_id', 'message_id'))
+    for (user_id, message_id) in notifications_to_update:
+        messages_by_user[user_id].append(message_id)
 
     num_detached = settings.MAX_UNBATCHED_REMOVE_NOTIFICATIONS - 1
     for user_profile_id in user_profile_ids:
+        filtered_message_ids = messages_by_user[user_profile_id]
         for message_id in filtered_message_ids[:num_detached]:
             # Older clients (all clients older than 2019-02-13) will only
             # see the first message ID in a given notification-message.
@@ -4201,8 +4209,7 @@ def do_clear_mobile_push_notifications_for_ids(user_profile_ids: List[int],
                 "user_profile_id": user_profile_id,
                 "message_ids": [message_id],
             })
-    if filtered_message_ids[num_detached:]:
-        for user_profile_id in user_profile_ids:
+        if filtered_message_ids[num_detached:]:
             queue_json_publish("missedmessage_mobile_notifications", {
                 "type": "remove",
                 "user_profile_id": user_profile_id,

```

Reply:
Oh! So this is what you meant, I was a little confused.